File: PHP/src/OneLang/Test/TestCaseRunner.php
<?php

namespace OneLang\Test\TestCaseRunner;

use OneLang\Test\TestCase\ITestCollection;
use OneLang\Test\TestCase\TestCase;
use OneLang\Test\TestCases\BasicTests\BasicTests;
use OneLang\Test\TestCases\ProjectGeneratorTest\ProjectGeneratorTest;

class TestResult {
    public $collectionName;
    public $testName;
    public $result;
    public $error;
    
    function __construct($collectionName, $testName, $result, $error) {
        $this->collectionName = $collectionName;
        $this->testName = $testName;
        $this->result = $result;
        $this->error = $error;
    }
}

class TestCaseRunner {
    public $tests;
    public $baseDir;
    
    function __construct($baseDir) {
        $this->baseDir = $baseDir;
        $this->tests = array();
        $this->tests[] = new BasicTests();
        $this->tests[] = new ProjectGeneratorTest($this->baseDir);
    }
    
    function runTests() {
        $results = array();
        \OneLang\Core\console::log("### TestCaseRunner -> START ###");
        
        foreach ($this->tests as $coll) {
            \OneLang\Core\console::log("### Collection -> " . $coll->name . " -> START ###");
            foreach ($coll->getTestCases() as $test) {
                \OneLang\Core\console::log("### TestCase -> " . $test->name . " -> START ###");
                try {
                    $artifactDir = $this->baseDir . "/test/artifacts/" . $coll->name . "/" . $test->name . "/" . "PHP" . "/";
                    $result = call_user_func($test->action, $artifactDir);
                    \OneLang\Core\console::log("### TestCase -> " . $test->name . " -> RESULT: " . json_encode($result, JSON_UNESCAPED_SLASHES) . " ###");
                    $results[] = new TestResult($coll->name, $test->name, $result, null);
                } catch (Exception $e) {
                    \OneLang\Core\console::log("### TestCase -> " . $test->name . " -> ERROR: " . $e . " ###");
                    $results[] = new TestResult($coll->name, $test->name, null, ($e)->message);
                }
                \OneLang\Core\console::log("### TestCase -> " . $test->name . " -> END ###");
            }
            \OneLang\Core\console::log("### Collection -> " . $coll->name . " -> END ###");
        }
        
        \OneLang\Core\console::log("### TestCaseRunner -> SAVE ###");
        OneFile::writeText($this->baseDir . "/test/artifacts/TestCaseRunner_results.json", json_encode($results, JSON_UNESCAPED_SLASHES));
        
        \OneLang\Core\console::log("### TestCaseRunner -> END ###");
        return $results;
    }
}

File: Python/OneLang/Test/TestCaseRunner.py
from onelang_core import *
import OneLang.index as index
import OneLang.Test.TestCase as testCase
import OneLang.Test.TestCases.BasicTests as basicTests
import OneLang.Test.TestCases.ProjectGeneratorTest as projGenTest
json

class TestResult:
    def __init__(self, collection_name, test_name, result, error):
        self.collection_name = collection_name
        self.test_name = test_name
        self.result = result
        self.error = error

class TestCaseRunner:
    def __init__(self, base_dir):
        self.tests = []
        self.base_dir = base_dir
        self.tests.append(basicTests.BasicTests())
        self.tests.append(projGenTest.ProjectGeneratorTest(self.base_dir))
    
    def run_tests(self):
        results = []
        console.log(f'''### TestCaseRunner -> START ###''')
        
        for coll in self.tests:
            console.log(f'''### Collection -> {coll.name} -> START ###''')
            for test in coll.get_test_cases():
                console.log(f'''### TestCase -> {test.name} -> START ###''')
                try:
                    artifact_dir = f'''{self.base_dir}/test/artifacts/{coll.name}/{test.name}/{"Python"}/'''
                    result = test.action(artifact_dir)
                    console.log(f'''### TestCase -> {test.name} -> RESULT: {json.dumps(result)} ###''')
                    results.append(TestResult(coll.name, test.name, result, None))
                except Exception as e:
                    console.log(f'''### TestCase -> {test.name} -> ERROR: {e} ###''')
                    results.append(TestResult(coll.name, test.name, None, (e).message))
                console.log(f'''### TestCase -> {test.name} -> END ###''')
            console.log(f'''### Collection -> {coll.name} -> END ###''')
        
        console.log(f'''### TestCaseRunner -> SAVE ###''')
        index.OneFile.write_text(f'''{self.base_dir}/test/artifacts/TestCaseRunner_results.json''', json.dumps(results))
        
        console.log(f'''### TestCaseRunner -> END ###''')
        return results